projects/plugins/jetpack/modules/widget-visibility/editor/index.jsx

```diff
@@ -0,0 +1,372 @@
+ */
+const maybeAddDefaultConditions = conditions => ( {
+	action: 'show',
+	rules: [],
+	match_all: 0,
+	...conditions,
+} );
+
+const visibilityAdvancedControls = createHigherOrderComponent(
+	BlockEdit => props => {
+		const { attributes, setAttributes, isSelected } = props;
+		const conditions = useMemo( () => attributes.conditions || {}, [ attributes ] );
+		const rules = useMemo( () => conditions.rules || [], [ conditions ] );
+
+		const toggleMatchAll = useCallback(
+			() =>
+				setAttributes( {
+					conditions: {
+						...maybeAddDefaultConditions( conditions ),
+						match_all: conditions.match_all === '0' ? '1' : '0',
+					},
+				} ),
+			[ setAttributes, conditions ]
+		);
+
+		const setAction = useCallback(
+			value =>
+				setAttributes( {
+					conditions: {
+						...maybeAddDefaultConditions( conditions ),
+						action: value,
+					},
+				} ),
+			[ setAttributes, conditions ]
+		);
+		const addNewRule = useCallback( () => {
+			const newRules = [ ...rules, { major: '', minor: '' } ];
+			setAttributes( {
+				conditions: {
+					...maybeAddDefaultConditions( conditions ),
+					rules: newRules,
+				},
+			} );
+		}, [ setAttributes, conditions, rules ] );
+
+		const deleteRule = useCallback(
+			i => {
+				const newRules = [ ...rules.slice( 0, i ), ...rules.slice( i + 1 ) ];
+				setAttributes( {
+					conditions: {
+						...maybeAddDefaultConditions( conditions ),
+						rules: newRules,
+					},
+				} );
+			},
+			[ setAttributes, conditions, rules ]
+		);
+
+		const setMajor = useCallback(
+			( i, majorValue ) => {
+				// When changing majors, also change the minor to the first available option
+				let minorValue = '';
+				if (
+					majorValue in widget_conditions_data &&
+					Array.isArray( widget_conditions_data[ majorValue ] ) &&
+					widget_conditions_data[ majorValue ].length > 0
+				) {
+					minorValue = widget_conditions_data[ majorValue ][ 0 ][ 0 ];
+				}
+
+				const newRules = [
+					...rules.slice( 0, i ),
+					{ ...rules[ i ], major: majorValue, minor: minorValue },
+					...rules.slice( i + 1 ),
+				];
+				setAttributes( {
+					conditions: {
+						...maybeAddDefaultConditions( conditions ),
+						rules: newRules,
+					},
+				} );
+			},
+			[ setAttributes, conditions, rules ]
+		);
+
+		const setMinor = useCallback(
+			( i, value ) => {
+				// Don't allow section headings to be set
+				if ( value && value.includes( '__HEADER__' ) ) {
+					return;
+				}
+				const newRules = [
+					...rules.slice( 0, i ),
+					{ ...rules[ i ], minor: value },
+					...rules.slice( i + 1 ),
+				];
+				setAttributes( {
+					conditions: {
+						...maybeAddDefaultConditions( conditions ),
+						rules: newRules,
+					},
+				} );
+			},
+			[ setAttributes, conditions, rules ]
+		);
+
+		let mainRender = null;
+		if ( rules.length === 0 ) {
+			mainRender = (
+				<BaseControl
+					id="widget-vis__wrapper"
```

Comment:
Mmm, there seems to be something amiss in how this is structured because the button should not need a label. Probably the right pattern to follow here is https://developer.wordpress.org/block-editor/reference-guides/components/base-control/#basecontrol-visuallabel